rethinkdb/datadog_checks/rethinkdb/document_db/query.py

```diff
@@ -0,0 +1,118 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from typing import Any, Callable, Iterator, List, Mapping, Sequence, Tuple
+
+from datadog_checks.base import AgentCheck
+from datadog_checks.base.log import CheckLoggingAdapter
+
+from .types import Enumeration, Group, Metric, MetricSpec, Modifier
+from .utils import dotted_join, lookup_dotted, to_time_elapsed
+
+
+class DocumentQuery(object):
+    """
+    A helper for retrieving metrics from document-oriented ("JSON") databases.
+    """
+
+    def __init__(
+        self,
+        source,  # type: Callable[..., Iterator[Tuple[Any, List[str]]]]
+        name,  # type: str
+        prefix,  # type: str
+        metrics=None,  # type: List[MetricSpec]
+        enumerations=None,  # type: List[Enumeration]
+        groups=None,  # type: List[Group]
+    ):
+        self.source = source
+        self.name = name
+        self.prefix = prefix
+        self.metrics = [] if metrics is None else metrics
+        self.enumerations = [] if enumerations is None else enumerations
+        self.groups = [] if groups is None else groups
+
+    def _make_metric_from_spec(self, document, spec, tags, logger):
```

Comment:
I think it would be better to pass/set the logger on `DocumentQuery` than passing it around as function param.

Same for the tags.

Reply:
Can't work:

- `DocumentQuery` objects are typically static (here they're defined statically in the `queries.py` file), which is good because it means they represent "executable" objects that can be called in various contexts.
- OTOH the logger and tags are only available at runtime (either from the check instance or the results of RethinkDB queries).